File: src/courses/data/catalog/02-dsa/dsa-core/implementations/cpp/backtracking.cpp
/**
 * C++ Implementation: Backtracking
 * Invariant: Pruned state space exploration, O(b^d)
 */
#include <iostream>
#include <vector>
#include <stdexcept>

template <typename T>
class Backtracking {
private:
    std::vector<T> elements;
public:
    // TODO: Implement primary invariant operations
    void push(const T& val) {
        elements.push_back(val);
    }
    size_t size() const { return elements.size(); }
};

File: src/courses/data/catalog/02-dsa/dsa-core/implementations/java/backtracking.java
/**
 * Java Implementation: Backtracking
 * Invariant: Pruned state space exploration, O(b^d)
 */
package seed.dsa.core;

import java.util.*;

public class Backtracking<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}
